src/flb_fstore.c

```diff
@@ -337,10 +478,21 @@ struct flb_fstore_stream *flb_fstore_stream_create(struct flb_fstore *fs,
     return fs_stream;
 }
 
-void flb_fstore_stream_destroy(struct flb_fstore_stream *stream, int delete)
-{
+void flb_fstore_stream_destroy(struct flb_fstore *fs,
+                               struct flb_fstore_stream *stream,
+                               int delete)
+{   
+    struct flb_fstore_file *fsf;
+    struct mk_list *head;
+
     if (delete == FLB_TRUE) {
         cio_stream_delete(stream->stream);
+
+        /* remove stream files from fstore files_up up list */
+        mk_list_foreach(head, &stream->files) {
```

Comment:
since the entry `fsf` will be removed from the list, it will generate a corruption with other linked elements, you have you use `mk_list_foreach_safe()` instead `mk_list_foreach()`

Reply:
files_up_remove removes file from `files_up` list in fstore. mk_list_foreach enumerates `files` list in stream. I think mk_list_foreach is okay in this case because files list not modified, though can replace with mk_list_foreach_safe()

https://github.com/fluent/fluent-bit/pull/4680/files/9b3fea6edf755786e2e80311020d5d3a829317d7#diff-d774a9676f644aabb1f4cd06bc9860f329e7505dbc6125bc3df5214448598e84R59

Reply:
Oh, is this because of multithreading, @edsiper? I didn't consider that.